snow/health.go

```diff
@@ -0,0 +1,130 @@
+// Copyright (C) 2025, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package snow
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/ava-labs/avalanchego/api/health"
+	"github.com/ava-labs/avalanchego/ids"
+	"github.com/ava-labs/avalanchego/utils/set"
+
+	"github.com/ava-labs/hypersdk/event"
+)
+
+const (
+	VMReadinessHealthChecker      = "snowVMReadiness"
+	UnresolvedBlocksHealthChecker = "snowUnresolvedBlocks"
+)
+
+func (v *VM[I, O, A]) RegisterHealthChecker(name string, healthChecker health.Checker) error {
+	if _, loaded := v.healthCheckers.LoadOrStore(name, healthChecker); loaded {
+		return fmt.Errorf("duplicate health checker for %s", name)
+	}
+
+	return nil
+}
+
+func (v *VM[I, O, A]) HealthChecker(name string) (interface{}, error) {
+	value, ok := v.healthCheckers.Load(name)
+	if !ok {
+		return nil, fmt.Errorf("health checker %s not found", name)
+	}
+
+	return value, nil
+}
+
+func (v *VM[I, O, A]) initHealthCheckers() error {
+	vmReadiness := NewVMReadiness(func() bool {
+		return v.ready
+	})
+	if err := v.RegisterHealthChecker(VMReadinessHealthChecker, vmReadiness); err != nil {
+		return err
+	}
+
+	unresolvedBlockHealthChecker := NewUnresolvedBlocksCheck[I]()
+	if err := v.RegisterHealthChecker(UnresolvedBlocksHealthChecker, unresolvedBlockHealthChecker); err != nil {
+		return err
+	}
+
+	// Subscribe to Reject event removing vacuously verified (unresolved) blocks
+	v.AddPreRejectedSub(unresolvedBlockHealthChecker.Resolve())
+
+	return nil
+}
+
+// VMReadiness is concrete health check that should mark itself ready if isReady function returns true
+type VMReadiness struct {
+	isReady func() bool
+}
+
+func NewVMReadiness(isReady func() bool) *VMReadiness {
+	return &VMReadiness{isReady: isReady}
+}
+
+func (v *VMReadiness) HealthCheck(_ context.Context) (interface{}, error) {
+	ready := v.isReady()
+	if !ready {
+		return nil, errors.New("vm is not ready")
```

Comment:
Should we return this map in both cases but change the value to `false` if it's not ready?